### ISSUE TITLE
Error Setting Rlimit  operation not permitted

### ISSUE THREAD

**Issue by kokizzu** [tockins/realize#23]
```
$ realize add
Error Setting Rlimit  operation not permitted
Realize - 1.1
A Go build system with file watchers, output streams and live reload. Run, build and watch file changes with custom paths
```

![image](https://cloud.githubusercontent.com/assets/1061610/19423663/cebee118-944c-11e6-881e-bebc0e12cc86.png)

```
$ realize run
Error Setting Rlimit  operation not permitted
Realize - 1.1
A Go build system with file watchers, output streams and live reload. Run, build and watch file changes with custom paths
```

![image](https://cloud.githubusercontent.com/assets/1061610/19423669/d8e97982-944c-11e6-9544-97385c50e528.png)

also empty window shown:

![image](https://cloud.githubusercontent.com/assets/1061610/19423691/119341f0-944d-11e6-9ec1-13064a607fcb.png)

if i use `--no-bin` it shows something:

![image](https://cloud.githubusercontent.com/assets/1061610/19423709/53d97872-944d-11e6-91b3-90683ad07092.png)

but it doesn't run normally like when using `codegangsta/gin`

```
gin -i -b /tmp/gin-bin # to run it on RAM instead of disk
```

![image](https://cloud.githubusercontent.com/assets/1061610/19423718/7058345c-944d-11e6-977d-7c6f3fcbddf2.png)

**Comment by Asoseil** [comment]
This is a problem of your system permissions generated by [syscall](https://golang.org/pkg/syscall/) library

**Comment by karlmutch** [comment]
More color for other who are confused by the above comment.

realize tries to set the open file limit using the syscall library and uses a value of 10,000 open files.  Typically this hits the hard limits of most virtual box and docker machines.  If you are dead keen on this tool then replacing the system call with something such as https://github.com/fsnotify/fsnotify would be the way to go but I am not sure if you pass a pull request as it is an additional dependency.

**Comment by Asoseil** [comment]
From fsnotify FAQ:
> There are OS-specific limits as to how many watches can be created
We already use fsnotify with syscall for increase these limits. On mac and linux the limit is around 200 files/directories #5

**Comment by geddski** [comment]
I just ran into this as well. Is there a workaround?

**Comment by pritambaral** [comment]
Increasing limits is a privileged operation on Linux, as it should. `realize` refuses to even start if it cannot increase the `NOFILE` limit. The error message doesn't even mention what the user can do in this case. Only upon reading the source is it obvious that the user has no way of telling `realize` to not try to increase limits or try a lower number.

The proper way would be to warn the user to increase the limits themself or suffer poor performance. As this stands, a freshly installed `realize` is completely unusable unless one runs it as root, which is a terrible idea anyway.

> On mac and linux the limit is around 200 files/directories

On both Arch Linux and Ubuntu, the limit of max watchable files is 8192 out of the box. The limit of max open files is 1024. Also, the two have no relation to each other. `inotify` is not limited by `NOFILE`; its limit is independent, and the [FAQ at fsnotify](https://github.com/fsnotify/fsnotify/wiki/FAQ) mentions that.

**TL;DR:**
1. Increasing NOFILE in a program instead of asking the user to do it themselves is unconventional
2. NOFILE has no relation with fsnotify limits
3. fsnotify limits are ridiculously high (8192) already

^On Linux anyway, though I doubt OS X would be too different.

**Comment by Asoseil** [comment]
We are working to add a specific field in the config file for manage independently this limit. On mac the limit is really 200.

**Comment by pritambaral** [comment]
> We are working to add a specific field in the config file for manage independently this limit. could be a good idea?

For UX? Only if you show a message describing what the user can do instead of exiting with just one vague line.

That won't fix the actual bug though.

Up in this thread, you cite #5. In that, the user had an NOFILE limit of 8192, and total no. files of only 3092. If you still hit `too many open files`, maybe the program is redundantly opening too many files or the user already had all those files open in at least two other programs.

In any case, this program *cannot* increase NOFILE limits without root. No matter what is set in the config, the OS will simply refuse with `operation not permitted`. The proper way would be to suggest the user to either edit `/etc/security/limits.conf` and re-login or to manually run `sudo sh -c 'ulimit -n 10000; su $SUDO_USER'` first and then run realize within that.

**Comment by Asoseil** [comment]
are you suggesting to raise in any case the method of realize that increases the file limit?

**Comment by Asoseil** [comment]
#25
